### DISCUSSION TITLE
Questions about custom plugins / buttons and graphs / charts

### DISCUSSION THREAD

**Discussion by ZehuaZhang** [facebook/lexical#5577] [Q&A]
I'm gradually migrate my current open source wysiwyg editor to lexical.

Could someone point me with right directions on
1) How to implement custom plugins / buttons?
2) Change editor behavior when detecting certain content, like link auto analyzer to show a media frame when the link is a media link, etc. What's the correct approach for this modification?
3) Implement charts / graphs of e.g. `mermaid`, `graphviz`, `echarts`, or `abc` music scores by change code snippets grammar type

**Comment by ivailop7** [comment]
You can have a look at the playground package in the repo and how decorator nodes are implemented for example for any widgets that is not text-based, and element nodes for things like mentions and others

**Reply by ZehuaZhang** [reply]
Thank you, as for question 3, do you have any example to demostrate to use external packages for say `mermaid`? @ivailop7  The way of using it is through inline code, like this

input
```
# ``` mermaid
# code line #1 with mermaid format to draw a graph
# code line #2
# code line #3 
# ```
```

output
mermaid graph

**Reply by ivailop7** [reply]
Read in the docs on node transforms, and see how to Markdown shortcuts plugin was implemented
